scipy/ndimage/_filters.py

```diff
@@ -47,7 +48,369 @@
+        axes = (axes,) if np.isscalar(axes) else axes
+    else:
+        axes = tuple(range(-n_axes, 0))
+
+    # If `origin` is provided, then it must be "broadcastable" to a tuple with length
+    # equal to the core dimensionality.
+    if origin is None:
+        origin = (0,) * n_axes
+    else:
+        origin = (origin,)*n_axes if np.isscalar(origin) else tuple(origin)
+        integral = [np.issubdtype(np.asarray(i).dtype, np.integer) for i in origin]
+        if not all(integral):
+            raise ValueError("All elements of `origin` must be integers.")
+        if not len(origin) == n_axes:
+            message = ("`origin` must be an integer or tuple of integers with length "
+                       "equal to the number of axes.")
+            raise ValueError(message)
+
+    # mode must be one of the allowed strings, and we should convert it to the
+    # value required by `np.pad` here.
+    valid_modes = {'reflect', 'constant', 'nearest', 'mirror', 'wrap',
+                   'grid-mirror', 'grid-constant', 'grid-wrap', 'valid'}
+    if mode not in valid_modes:
+        raise ValueError(f"`mode` must be one of {valid_modes}.")
+    mode_map = {'nearest': 'edge', 'reflect': 'symmetric', 'mirror': 'reflect',
+                'grid-mirror': 'reflect', 'grid-constant': 'constant',
+                'grid-wrap': 'wrap'}
+    mode = mode_map.get(mode, mode)
+
+    if mode == 'valid' and any(origin):
+        raise ValueError("`mode='valid'` is incompatible with use of `origin`.")
+
+    if cval is None:
+        cval = 0.0
+    elif mode != 'constant':
+        raise ValueError("Use of `cval` is compatible only with `mode='constant'`.")
+
+    # `cval` must be a scalar or "broadcastable" to a tuple with the same
+    # dimensionality of `input`. (Full input validation done by `np.pad`.)
+    if not np.issubdtype(np.asarray(cval).dtype, np.number):
+        raise ValueError("`cval` must include only numbers.")
+
+    # `batch_memory` must be a positive number.
+    temp = np.asarray(batch_memory)
+    if temp.ndim != 0 or (not np.issubdtype(temp.dtype, np.number)) or temp <= 0:
+        raise ValueError("`batch_memory` must be positive number.")
+
+    # For simplicity, work with `axes` at the end.
+    working_axes = tuple(range(-n_axes, 0))
+    if axes is not None:
+        input = np.moveaxis(input, axes, working_axes)
+        output = (np.moveaxis(output, axes, working_axes)
+                  if output is not None else output)
+
+    # Wrap the function to limit maximum memory usage, deal with `footprint`,
+    # and populate `output`. The latter requires some verbosity because we
+    # don't know the output dtype.
+    def function(view, output=output, function=function):
+        kwargs = {'axis': working_axes}
+
+        if working_axes == ():
+            return function(view, **kwargs)
+
+        # for now, assume we only have to iterate over zeroth axis
+        chunk_size = math.prod(view.shape[1:]) * view.dtype.itemsize
```

Comment:
Makes me lean even stronger towards specifying the memory_budget in terms of array elements not bytes.